File: apps/web/src/lib/api/applicant.ts
/**
 * Typed API functions for applicant-side endpoints.
 * Mirrors the Pydantic schemas in apps/api/app/schemas/applicant.py.
 */
import { apiFetch } from "./client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ApplicantProfileSummary {
  applicant_id: string;
  first_name: string | null;
  last_name: string | null;
  program_name_raw: string | null;
  canonical_job_family_code: string | null;
  city: string | null;
  state: string | null;
  region: string | null;
  /** Geocoded home coordinates (city-level centroid) — may be null. */
  lat: number | null;
  lng: number | null;
  willing_to_relocate: boolean;
  willing_to_travel: boolean;
  commute_radius_miles: number | null;
  expected_completion_date: string | null;
  available_from_date: string | null;
  profile_completeness: number;
}

export interface JobMatchSummary {
  match_id: string;
  job_id: string;
  job_title: string;
  employer_name: string;
  is_partner_employer: boolean;
  job_city: string | null;
  job_state: string | null;
  job_region: string | null;
  work_setting: string | null;
  travel_requirement: string | null;
  geography_note: string | null;
  pay_min: number | null;
  pay_max: number | null;
  pay_type: string | null;
  eligibility_status: "eligible" | "near_fit" | "ineligible";
  match_label: "strong_fit" | "good_fit" | "moderate_fit" | "low_fit" | null;
  policy_adjusted_score: number | null;
  top_strengths: string[];
  top_gaps: string[];
  recommended_next_step: string | null;
  source_url: string | null;
  canonical_job_family_code: string | null;
  /** In-platform apply available for this job — powers the list's Apply sheet. */
  internal_apply: boolean;
  description_raw: string | null;
  requirements_raw: string | null;
  preferred_qualifications_raw: string | null;
  experience_level: string | null;
  confidence_level: "high" | "medium" | "low" | null;
  requires_review: boolean;
  applicant_interest: "interested" | "applied" | "not_interested" | null;
  // Relaxation tier (visibility/grouping only — never a score change)
  match_tier: "strict" | "adjacent" | "stretch" | "nearby" | null;
  tier_reason: string | null;
  distance_miles: number | null;
}

export interface DimensionScoreItem {
  dimension: string;
  weight: number;
  raw_score: number;
  weighted_score: number;
  rationale: string | null;
  null_handling_applied: boolean;
  null_handling_default: number | null;
}

export interface GateResultItem {
  gate_name: string;
  result: "pass" | "near_fit" | "fail";
  reason: string;
  severity: string | null;
}

export interface PolicyModifierItem {
  policy: string;
  value: number;
  reason: string;
}

export interface JobMatchDetail extends JobMatchSummary {
  base_fit_score: number | null;
  weighted_structured_score: number | null;
  semantic_score: number | null;
  required_missing_items: string[];
  hard_gate_rationale: GateResultItem[];
  policy_modifiers: PolicyModifierItem[];
  dimension_scores: DimensionScoreItem[];
}

export interface RankedMatchesResponse {
  applicant_id: string;
  eligible_matches: JobMatchSummary[];
  total_eligible: number;
  near_fit_matches: JobMatchSummary[];
  total_near_fit: number;
  // "Near you, different trade": geography-only relaxation tier, populated
  // only when the stricter sections fall below the configured floor.
  nearby_matches: JobMatchSummary[];
  total_nearby: number;
  relaxation_applied: boolean;
  has_matches: boolean;
  profile_has_family: boolean;
  profile_has_location: boolean;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function fetchMyProfile(
  token: string,
): Promise<ApplicantProfileSummary> {
  return apiFetch<ApplicantProfileSummary>("/applicant/me/profile", token);
}

/** Per-tier pagination for the ranked-matches endpoint (all additive). */
export interface MatchesPageOpts {
  eligibleOffset?: number;
  nearFitOffset?: number;
  nearbyOffset?: number;
  limit?: number;
}

export async function fetchMyMatches(
  token: string,
  opts?: MatchesPageOpts,
): Promise<RankedMatchesResponse> {
  const qs = new URLSearchParams();
  if (opts?.eligibleOffset) qs.set("eligible_offset", String(opts.eligibleOffset));
  if (opts?.nearFitOffset) qs.set("near_fit_offset", String(opts.nearFitOffset));
  if (opts?.nearbyOffset) qs.set("nearby_offset", String(opts.nearbyOffset));
  if (opts?.limit) qs.set("limit", String(opts.limit));
  const suffix = qs.size > 0 ? `?${qs.toString()}` : "";
  return apiFetch<RankedMatchesResponse>(`/applicant/me/matches${suffix}`, token);
}

/** Achievement badges — every one computed from real recorded activity. */
export interface ApplicantBadge {
  key: string;
  title: string;
  description: string;
  earned: boolean;
  earned_at: string | null;
  progress: { current: number; target: number };
}

export interface ApplicantBadgesResponse {
  badges: ApplicantBadge[];
  earned_count: number;
  total_count: number;
}

export async function fetchMyBadges(
  token: string,
): Promise<ApplicantBadgesResponse> {
  return apiFetch<ApplicantBadgesResponse>("/applicant/me/badges", token);
}

export async function fetchMatchDetail(
  matchId: string,
  token: string,
): Promise<JobMatchDetail> {
  return apiFetch<JobMatchDetail>(`/applicant/me/matches/${matchId}`, token);
}

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

export function formatWorkSetting(ws: string | null): string {
  switch (ws) {
    case "remote":    return "Remote";
    case "hybrid":    return "Hybrid";
    case "on_site":   return "On-site";
    case "flexible":  return "Flexible";
    default:          return ws ?? "—";
  }
}

export function formatPay(
  min: number | null,
  max: number | null,
  type: string | null,
): string {
  if (min === null) return "Pay not listed";
  const suffix = type === "hourly" ? "/hr" : type === "annual" ? "/yr" : "";
  const fmt = (n: number) =>
    type === "annual"
      ? `$${(n / 1000).toFixed(0)}k`
      : `$${n.toFixed(0)}`;
  if (max && max !== min) return `${fmt(min)}–${fmt(max)}${suffix}`;
  return `${fmt(min)}${suffix}`;
}

/**
 * Suppress the "~0 mi away" clause in engine-written strength/gap strings.
 * Matches CommuteChip's documented rule: distances under a mile are noise —
 * "In your city" already says it, and "~0 mi away" reads as a bug.
 */
export function stripZeroDistance(text: string): string {
  return text
    .replace(/\s*[—–,-]\s*~0 mi away\s*/g, " ")
    .replace(/ {2,}/g, " ")
    .replace(/\s+([,.;])/g, "$1")
    .trim();
}

export function formatDimensionName(dim: string): string {
  return dim
    .split("_")
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(" ");
}

File: apps/api/app/schemas/applicant.py
"""
Pydantic response schemas for applicant-side API endpoints (Phase 6.1).

Key design rules (DECISIONS.md / SCORING_CONFIG.yaml):
  - Display score: policy_adjusted_score (per §ui_visibility)
  - base_fit_score stored separately, shown in detail view only
  - Two labeled sections: eligible ("best immediate") / near_fit ("promising near-fit")
  - Ineligible matches hidden from applicant by default
  - Geography is first-class in every response
"""
from __future__ import annotations
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Profile
# ---------------------------------------------------------------------------

class ApplicantProfileSummary(BaseModel):
    applicant_id: str
    first_name: str | None
    last_name: str | None
    program_name_raw: str | None
    canonical_job_family_code: str | None
    city: str | None
    state: str | None
    region: str | None
    willing_to_relocate: bool
    willing_to_travel: bool
    expected_completion_date: str | None
    available_from_date: str | None
    profile_completeness: int   # 0–100 rough percentage

    # Expanded fields
    enrollment_status: str | None
    degree_type: str | None
    school_name: str | None
    school_city: str | None
    school_state: str | None
    career_path: str | None
    program_field: str | None
    specific_career: str | None
    program_start_date: str | None
    gpa: float | None

    travel_preference: str | None
    relocation_preference: str | None
    relocation_states: list[str]

    age_range: str | None
    gender: str | None
    military_status: bool
    military_dependent: bool
    current_wages: str | None
    has_internship: bool
    activities: str | None
    honor_societies: list[str]


# ---------------------------------------------------------------------------
# Match list items
# ---------------------------------------------------------------------------

class JobMatchSummary(BaseModel):
    match_id: str
    job_id: str
    job_title: str                  # title_normalized or title_raw fallback
    employer_name: str
    is_partner_employer: bool

    # Geography (first-class)
    job_city: str | None
    job_state: str | None
    job_region: str | None
    work_setting: str | None        # remote / hybrid / on_site / flexible
    travel_requirement: str | None
    geography_note: str | None      # derived human-readable note

    # Pay
    pay_min: float | None
    pay_max: float | None
    pay_type: str | None            # hourly / annual

    # Score and label (DECISIONS.md §ui: display policy_adjusted_score)
    eligibility_status: str         # eligible / near_fit
    match_label: str | None         # strong_fit / good_fit / moderate_fit / low_fit
    policy_adjusted_score: float | None

    # Explanation
    top_strengths: list[str]
    top_gaps: list[str]
    recommended_next_step: str | None

    # Application
    source_url: str | None
    canonical_job_family_code: str | None

    # Job detail (for expandable card)
    description_raw: str | None
    requirements_raw: str | None
    preferred_qualifications_raw: str | None
    experience_level: str | None

    # Confidence
    confidence_level: str | None
    requires_review: bool


# ---------------------------------------------------------------------------
# Match detail (extends summary)
# ---------------------------------------------------------------------------

class DimensionScoreItem(BaseModel):
    dimension: str
    weight: float
    raw_score: float
    weighted_score: float
    rationale: str | None
    null_handling_applied: bool
    null_handling_default: float | None


class GateResultItem(BaseModel):
    gate_name: str
    result: str          # pass / near_fit / fail
    reason: str
    severity: str | None


class PolicyModifierItem(BaseModel):
    policy: str
    value: float
    reason: str


class JobMatchDetail(JobMatchSummary):
    # Additional score layers (detail view only)
    base_fit_score: float | None
    weighted_structured_score: float | None
    semantic_score: float | None

    # Mandatory gaps (hard failures)
    required_missing_items: list[str]

    # Gate breakdown
    hard_gate_rationale: list[GateResultItem]

    # Policy adjustments
    policy_modifiers: list[PolicyModifierItem]

    # Dimension scores (9 rows)
    dimension_scores: list[DimensionScoreItem]


# ---------------------------------------------------------------------------
# Ranked matches response (two-section applicant view)
# ---------------------------------------------------------------------------

class RankedMatchesResponse(BaseModel):
    applicant_id: str

    # Section 1: "Best immediate opportunities" — eligible only
    eligible_matches: list[JobMatchSummary]
    total_eligible: int

    # Section 2: "Promising near-fit opportunities" — near_fit only
    near_fit_matches: list[JobMatchSummary]
    total_near_fit: int

    has_matches: bool
    profile_has_family: bool    # job family normalised
    profile_has_location: bool  # state / region set
